### PR TITLE
Add document of uuid filter and schema for secret data for Onvif Discovery Handler

### PR DESCRIPTION
**Pull request by johnsonshih**
Add uuid filter document
Add document of secret data schema used by Onvif Discovery Handler
Add document of secret data schema used by Onvif sample broker

### CODE REVIEW

**Review by diconico07 — COMMENTED** [review]
Took a first look at this, looks quite good to me, just noticed a few minor issues.

File: docs/discovery-handlers/onvif.md [hunk 1]
```diff
@@ -34,8 +34,14 @@ Discovery Handlers are passed discovery details that are set in a Configuration
 | onvif.configuration.discoveryDetails.macAddresses.items | array of mac addresses | empty | mac addresses that the filter action acts upon |
 | onvif.configuration.discoveryDetails.scope.action | Include, Exclude | Exclude | filter action to take on a set of scopes |
 | onvif.configuration.discoveryDetails.scope.items | array of scopes | empty | scopes that the filter action acts upon |
+| onvif.configuration.discoveryDetails.uuids.action* | Include, Exclude | Exclude | filter action to take on a set of device uuids |
+| onvif.configuration.discoveryDetails.uuids.items* | array of IP addresses | empty | device uuids that the filter action acts upon |
```
[the remark this answers is not in the document]
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:38]: correct typo

File: docs/discovery-handlers/onvif.md [hunk 2]
```diff
@@ -88,6 +94,195 @@ By default, if a broker Pod is specified, a single broker Pod is deployed to eac
...
 
+### Discovery Handler Discovery Properties Settings
+
+Agent read the the content of `discoveryProperties` in Configuration and generate a string key-value pair list to Discovery Handler.  The Onvif discovery handler
+leverage the `discoveryProperties` to read the credentials for authenticated discovery. There are two attributes required for Onvif discovery handler to perform
+authenticated discovery:
+1. an id that can unique identify a camera
+2. a credential (username/password) to authenticate the access to a camera
+
+Onvif discovery handler can get the device uuid when discovering Onvif camera devices, and use the id to look up matched credential from the string key-value pair list passed by Agent.
```
[the remark this answers is not in the document]
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:105]: adjust the statement

File: docs/discovery-handlers/onvif.md [hunk 3]
```diff
@@ -88,6 +94,195 @@ By default, if a broker Pod is specified, a single broker Pod is deployed to eac
...
+3.	Device credential entry
+
+All three ways can be used in the same Configuration, the order above is the order of Onvif Discovery Handler processing the secret 
+information. If there is any secret information duplication between different groups, the latter overwrites the prior entries.
+If there is any duplication within the same group, it’s up to the Onvif Discovery Handler to decide which one wins when processing
+the entries, and it’s not guaranteed the order is always the same.
+
+##### Device credential list
+Onvif Discovery Handler first looks for a key named “`device_credential_list`” and expects its value points to a string array of credential
+lists in json format.  The value is an array of credential list which points to a Kubernetes secret, the secret is expected in json format.
```
[the remark this answers is not in the document]
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:124]: remove

File: docs/discovery-handlers/onvif.md [hunk 4]
```diff
@@ -88,6 +94,192 @@ By default, if a broker Pod is specified, a single broker Pod is deployed to eac
 | :--- | :--- | :--- | :--- |
 | onvif.configuration.capacity | number | 1 | maximum number of brokers that can be deployed to utilize a device (up to 1 per Node) |
 
+### Discovery Handler Discovery Properties Settings
+
+Agent read the the content of `discoveryProperties` in Configuration and generate a string key-value pair list to Discovery Handler.  The Onvif discovery handler
```
Comment (diconico07) [inline docs/discovery-handlers/onvif.md:99]:
```suggestion
Agent read the content of `discoveryProperties` in Configuration and generate a string key-value pair list to Discovery Handler.  The Onvif discovery handler
```

**Review by diconico07 — APPROVED** [review]
LGTM, just noticed another quick typo.

File: docs/discovery-handlers/onvif.md (hunk 4 above)
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md]: fixed

File: docs/discovery-handlers/onvif.md [hunk 5]
```diff
@@ -88,6 +94,192 @@ By default, if a broker Pod is specified, a single broker Pod is deployed to eac
...
+    - name: "device_credential_ref_list"
+      value: |+
+        [
+          "secret_ref_list1",
+          "secret_ref_list2"
+        ]
+    - name: "secret_ref_list1"
+      value: |+
+        {
+          "5f5a69c2-e0ae-504f-829b-00fcdab169cc" : { "username_ref" : "device1_username", "password_ref" : "device1_password" },
```
Comment (bfjelds) [inline docs/discovery-handlers/onvif.md:188]: why sometimes username_ref and sometimes username?

Comment (bfjelds) [reply docs/discovery-handlers/onvif.md:188]: nevermind.  thought i was looking at something else.

File: docs/discovery-handlers/onvif.md [hunk 6]
```diff
@@ -156,6 +361,65 @@ helm install akri akri-helm-charts/akri \
     --set onvif.configuration.discoveryDetails.discoveryTimeoutSeconds=2
 ```
 
+### Accessing Secret data in Broker
+
+The Onvif a sample broker (`akri-onvif-video-broker`) can be configured to access Secret and ConfigMap data, if configured, it expects the Secret and ConfigMap data are
```
Comment (bfjelds) [inline docs/discovery-handlers/onvif.md:366]: nit: remove `a` ... `The Onvif sample broker...`

File: docs/discovery-handlers/onvif.md [hunk 7]
```diff
@@ -156,6 +361,65 @@ helm install akri akri-helm-charts/akri \
...
+exists, it tries to get the same file name under `CREDENTIAL_DIRECTORY`.  The credential ref list should contains reference entries to the actual credentals. The sample
+broker look up the device uuid from the list to get the credential referece and read the actual credential files from `CREDENTIAL_DIRECTORY`.  Similar to the fallback credenial "username_default" and "password_default", a credential ref entry with key "default" indicates the fallback credental.
+3. if the sample broker can not find a matched credental from credential ref list, it looks for credentials from crdential list `device_credential_list`.  The sample broker try to get a file name `device_credential_list` from `CREDENTIAL_CONFIGMAP_DIRECTORY`, and if the file does not xxists, it tries to get the same file name under `CREDENTIAL_DIRECTORY`.  The credential list should contains actual credential entries. The sample broker look up the device uuid from the list to get the credential. Similarly, a credential entry with key "default" indicates the fallback credental.
```
Comment (bfjelds) [inline docs/discovery-handlers/onvif.md:378]: nit: `xxists` should be `exists`

File: docs/discovery-handlers/onvif.md [hunk 8]
```diff
@@ -156,6 +361,65 @@ helm install akri akri-helm-charts/akri \
...
+metadata:
+  name: onvif-auth-configmap
+data:
+  device_credential_ref_list: |+ 
+    [ "credential_ref_list1" ]
+  credential_ref_list1: |+
+    {
+        "6a67158b-42b1-400b-8afe-1bec9a5d7919":
+            {
+                "username_ref" : "device2_username",
```
Comment (bfjelds) [inline docs/discovery-handlers/onvif.md:395]: should device2_username just be username?

Comment (bfjelds) [reply docs/discovery-handlers/onvif.md:395]: (same for password)

Comment (bfjelds) [reply docs/discovery-handlers/onvif.md:395]: i'm probably just missing how device2_username connects to username_6a67158b_42b1_400b_8afe_1bec9a5d7909.  maybe they don't?

Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:395]: good catch, device2_username and device2_password are missing, I'll add them back

File: docs/discovery-handlers/onvif.md (hunk 6 above)
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:366]: fix in https://github.com/project-akri/akri-docs/pull/89

File: docs/discovery-handlers/onvif.md (hunk 7 above)
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:378]: fix in https://github.com/project-akri/akri-docs/pull/89

File: docs/discovery-handlers/onvif.md (hunk 8 above)
Comment (johnsonshih) [reply docs/discovery-handlers/onvif.md:395]: fix in https://github.com/project-akri/akri-docs/pull/89

**Pull request merged**